File: dart/lib/src/parser/parser.dart
import 'dart:io';

import 'package:dart_deez/src/lexer/lexer.dart';
import 'ast.dart';

export 'ast.dart';

String red(String s) => "\x1b[31m$s\x1b[0m";

Never parseError(String message) {
  print(message);
  print(StackTrace.current);
  exit(1);
}

class Parser {
  final Lexer lexer;
  final errors = <String>[];
  // late implies that the variable will be initialized but not directly in the constructor.
  late Token _currentToken = const Token.eof();
  late Token _peekToken = const Token.eof();

  Parser(this.lexer) {
    _nextToken();
    _nextToken();
  }

  Program parseProgram() {
    final statements = <Statement>[];

    while (_currentToken.type != TokenType.eof) {
      final statement = _parseStatement();
      if (statement != null) {
        statements.add(statement);
      }
      _nextToken();
    }

    return Program(statements);
  }

  Statement? _parseStatement() {
    return switch (_currentToken.type) {
      TokenType.let => _parseLetStatement(),
      TokenType.return_ => _parseReturnStatement(),
      _ => null,
    };
  }

  Statement? _parseLetStatement() {
    final token = _currentToken;

    if (!_expectPeek(TokenType.ident)) {
      return null;
    }

    final name = Identifier(_currentToken, _currentToken.literal);

    if (!_expectPeek(TokenType.assign)) {
      return null;
    }

    while (!_currentTokenIs(TokenType.semicolon)) {
      _nextToken();
    }

    return LetStatement(
      token,
      name,
      Identifier(_currentToken, _currentToken.literal),
    );
  }

  Statement? _parseReturnStatement() {
    final token = _currentToken;

    _nextToken();

    while (!_currentTokenIs(TokenType.semicolon)) {
      _nextToken();
    }

    return ReturnStatement(
      token,
      Identifier(_currentToken, _currentToken.literal),
    );
  }

  bool _currentTokenIs(TokenType type) {
    return _currentToken.type == type;
  }

  bool _peekTokenIs(TokenType type) {
    return _peekToken.type == type;
  }

  void _nextToken() {
    _currentToken = _peekToken;
    _peekToken = lexer.nextToken();
  }

  bool _expectPeek(TokenType type) {
    if (_peekToken.type == type) {
      _nextToken();
      return true;
    } else {
      _peekError(type);
      return false;
    }
  }

  void _peekError(TokenType type) {
    errors.add(
      red("Expected next token to be ${type.name}, got ${_peekToken.type.name}"),
    );
  }

  bool checkParserErrors() {
    if (errors.isEmpty) {
      return false;
    }
    print(errors.join("\n"));
    return true;
  }
}

File: java/monkey/src/main/java/root/Parser.java
package root;

import root.ast.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Parser {

    private final Lexer lexer;

    private Token currentToken;

    private Token peekToken;

    public final List<String> errors = new ArrayList<>();

    public Parser(Lexer lexer) {
        this.lexer = lexer;

        this.nextToken();
        this.nextToken();
    }

    public Program parseProgram() {
        Program program = new Program();

        while (!curTokenIs(TokenType.EOF)) {
            parseStatement().ifPresent(program.statements::add);
            nextToken();;
        }

        return program;
    }

    private void peekError(TokenType expected) {
        errors.add("Expected next token to be %s, got %s".formatted(expected.name(), currentToken.type().name()));
    }

    private void nextToken() {
        this.currentToken = this.peekToken;
        this.peekToken = this.lexer.nextToken();
    }

    private Optional<Statement> parseStatement() {
        return switch (this.currentToken.type()) {
            case LET -> this.parseLetStatement();
            case RETURN -> this.parseReturnStatement();
            default -> Optional.empty();
        };
    }

    private Optional<Statement> parseReturnStatement() {
        ReturnStatement returnStatement = new ReturnStatement(this.currentToken);

        this.nextToken();

        // TODO: We're skipping the expressions until we encounter a semicolon
        while (!this.curTokenIs(TokenType.SEMI)) {
            this.nextToken();
        }

        return Optional.of(returnStatement);
    }

    private Optional<Statement> parseLetStatement() {
        LetStatement letStatement = new LetStatement(this.currentToken);

        if (!this.expectPeek(TokenType.IDENT)) {
            return Optional.empty();
        }

        letStatement.name = new Identifier(this.currentToken, this.currentToken.literal());

        if (!this.expectPeek(TokenType.ASSIGN)) {
            return Optional.empty();
        }

        // TODO: We're skipping the expressions until we encounter a semicolon
        while (!this.curTokenIs(TokenType.SEMI)) {
            this.nextToken();
        }

        return Optional.of(letStatement);
    }
    private boolean curTokenIs(TokenType type) {
        return this.currentToken.type() == type;
    }

    private boolean peekTokenIs(TokenType type) {
        return this.peekToken.type() == type;
    }

    private boolean expectPeek(TokenType type) {
        if (this.peekTokenIs(type)) {
            this.nextToken();
            return true;
        }

        this.peekError(type);
        return false;
    }
}
